metricbeat/module/iis/application_pool/reader.go

```diff
@@ -0,0 +1,276 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build windows
+
+package application_pool
+
+import (
+	"strings"
+
+	"github.com/elastic/go-sysinfo"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/helper/windows/pdh"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+// Reader  strucr will contain the pdh query and config options
+type Reader struct {
+	Query            pdh.Query         // PDH Query
+	ApplicationPools []ApplicationPool // Mapping of counter path to key used for the label (e.g. processor.name)
+	log              *logp.Logger      // logger
+	hasRun           bool              // will check if the reader has run a first time
+	WorkerProcesses  map[string]string
+}
+
+// ApplicationPool struct contains the list of applications and their worker processes
+type ApplicationPool struct {
+	Name             string
+	WorkerProcessIds []int
+	counters         map[string]string
+}
+
+// WorkerProcess struct contains the worker process details
+type WorkerProcess struct {
+	ProcessId    int
+	InstanceName string
+}
+
+const ecsProcessId = "process.pid"
+
+var appPoolCounters = map[string]string{
+	"process.pid":                          "\\Process(w3wp*)\\ID Process",
+	"process.cpu_usage_perc":               "\\Process(w3wp*)\\% Processor Time",
+	"process.handle_count":                 "\\Process(w3wp*)\\Handle Count",
+	"process.thread_count":                 "\\Process(w3wp*)\\Thread Count",
+	"process.working_set":                  "\\Process(w3wp*)\\Working Set",
+	"process.private_byte":                 "\\Process(w3wp*)\\Private Bytes",
+	"process.virtual_bytes":                "\\Process(w3wp*)\\Virtual Bytes",
+	"process.page_faults_per_sec":          "\\Process(w3wp*)\\Page Faults/sec",
+	"process.io_read_operations_per_sec":   "\\Process(w3wp*)\\IO Read Operations/sec",
+	"process.io_write_operations_per_sec":  "\\Process(w3wp*)\\IO Write Operations/sec",
+	"net_clr.total_exceptions_thrown":      "\\.NET CLR Exceptions(w3wp*)\\# of Exceps Thrown",
```

Comment:
Additional .NET CLR metrics:
* `.NET CLR Memory(w3wp)\% Time in GC`
The percentage of elapsed time that was spent in performing a garbage collection (GC) since the last GC cycle.
* `.NET CLR Memory(w3wp)\# Bytes in all Heaps` 
The sum of four other counters—Gen 0 Heap Size, Gen 1 Heap Size, Gen 2 Heap Size, and the Large Object Heap Size. 
* `.NET CLR Memory(w3wp)\# Total Committed Bytes`
The amount of virtual memory (in bytes) currently committed by the Garbage Collector. 
* `.NET CLR Memory(w3wp)\Process ID` 
The process ID of the CLR process instance being monitored.
* `.NET CLR Security(w3wp)\% Time in RT checks` 
The percentage of elapsed time spent in performing run-time Code Access Security (CAS) checks since the last such check. It represents the last observed value; it’s not an average.

Reply:
@sorantis , adding almost all the counters suggested in all 3 metricsets more than doubled the CPU usage, do you mind going over the lists again and picking just a few that could be of interest.

Reply:
skip